### ISSUE TITLE
Support false-progress fail-over scenarios

### ISSUE THREAD

**Issue by tmasternak** [Lockstitch/Lockstitch#7]
Secondary replicas can become subject of so called [false progress](https://github.com/Microsoft/azure-docs/blob/master/articles/service-fabric/service-fabric-reliable-services-notifications.md#recommendations). There are many scenarios in which that can happen - here is one to give an example. Let's assume we have replica set of size 5 and require 3 acks to consider operation to be committed:
 * [Epoch=n] Primary replicates operation `op-a` at log index `x` to all secondaries 
 * [Epoch=n] Only one secondary received `op-a`  acked it and failed
 * [Epoch=n] Primary fails 
 * Fail-over Service runs an election and one of the 3 remaining secondaries becomes new primary
 * [Epoch=n+1] New primary replicates successfully `op-b` at long index `x`
 * [Epoch=n+1] Failed secondary gets back on-line

Now we have a problem because new primary and old secondary have inconsistent log entries at index `x`. This need to be sorted out during secondary catch-up phase ([GetCopyState](https://github.com/Lockstitch/Lockstitch/blob/master/Lockstich/Lockstitch.Service/ReplicatedServiceBase.cs#L194) and [GetCopyContext](https://github.com/Lockstitch/Lockstitch/blob/master/Lockstich/Lockstitch.Service/ReplicatedServiceBase.cs#L187)).

### Solution
The secondary has to provide enough information in copy state to enable current primary to detect false progress and instruct the secondary to override some of the entries in it's log. 

An naive approach would be to send information about the content of the log but that would lead to potentially large amounts of data being sent. The alternative is to use progress vector that holds log information per each epoch replica participated in. Each entry in the vector is [`epoch-number`, `max-log-index`, `data-loss-count`] where `max-log-index` is the biggest log index seen by replica in the `epoch-number` and `data-loss-count` is the number of data-loss events replica knows about.

This should be enough for primary to identify log entry that is first to be copied to secondary that will handle false progress correctly.

**Comment by tmasternak** [comment]
@Scooletz this is the first stab at the proper replication protocol. The assumption is that the details of progress vector based replication will be filled in as this issue will start progressing.

Is the problem description clear?

**Comment by Scooletz** [comment]
LGTM @tmasternak

**Comment by Scooletz** [comment]
Is there a cut-off point where the vector can be truncated? Isn't it effectively only `one epoch` ahead?

**Comment by tmasternak** [comment]
That's a good question. For now I would say that we would have to introduce some additional operation ala  global commit index from Raft. 

Cutting off one epoch in the past will probably not be enough because we can have `InvalidSecondary` that went down for more than one epoch.

**Comment by Scooletz** [comment]
You're right. We could introduce a split in the very beginning of the log and that would ruin this approach. 
Don't we need some messaging in here then? Something that would be able to send a `heartbeat` with the `epoch/commitIndex` tuples? That would make it much easier to work with, wouldn't it?

**Comment by tmasternak** [comment]
This is getting interesting :). Here are my thoughts:
 * We could allow secondaries to do false progress - SF reliable collections are already doing that. But this introduces complexity to anything that runs on top of log because it needs to be able to undo operations that have been already applied.
 * We need some way to truncate/snapshot the log and be sure that snapshot is valid (does not contain any false progress). So in other words we might not need `epoch/commitIndex` sent frequently but we need it at least to coordinate safe points for doing log snapshots.

Considering both of those things I would propose to introduce `epoch/commitIndex` messages and later on we can discuss if removing them brings benefits in some scenarios.
